x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/TokenPruningConfig.java

```diff
@@ -0,0 +1,162 @@
+        this.tokensFreqRatioThreshold = in.readFloat();
+        this.tokensWeightThreshold = in.readFloat();
+        this.onlyScorePrunedTokens = in.readBoolean();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeFloat(tokensFreqRatioThreshold);
+        out.writeFloat(tokensWeightThreshold);
+        out.writeBoolean(onlyScorePrunedTokens);
+    }
+
+    /**
+     * Returns the frequency ratio threshold to apply on the query.
+     * Tokens whose frequency is more than ratio_threshold times the average frequency of all tokens in the specified
+     * field are considered outliers and may be subject to removal from the query.
+     */
+    public float getTokensFreqRatioThreshold() {
+        return tokensFreqRatioThreshold;
+    }
+
+    /**
+     * Returns the weight threshold to apply on the query.
+     * Tokens whose weight is more than (weightThreshold * best_weight) of the highest weight in the query are not
+     * considered outliers, even if their frequency exceeds the specified ratio_threshold.
+     * This threshold ensures that important tokens, as indicated by their weight, are retained in the query.
+     */
+    public float getTokensWeightThreshold() {
+        return tokensWeightThreshold;
+    }
+
+    /**
+     * Returns whether the filtering process retains tokens identified as non-relevant based on the specified thresholds
+     * (ratio and weight). When {@code true}, only non-relevant tokens are considered for matching and scoring documents.
+     * Enabling this option is valuable for re-scoring top hits retrieved from a {@link WeightedTokensQueryBuilder} with
+     * active thresholds.
+     */
+    public boolean isOnlyScorePrunedTokens() {
+        return onlyScorePrunedTokens;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        TokenPruningConfig that = (TokenPruningConfig) o;
+        return Float.compare(that.tokensFreqRatioThreshold, tokensFreqRatioThreshold) == 0
+            && Float.compare(that.tokensWeightThreshold, tokensWeightThreshold) == 0
+            && onlyScorePrunedTokens == that.onlyScorePrunedTokens;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(tokensFreqRatioThreshold, tokensWeightThreshold, onlyScorePrunedTokens);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(PRUNING_CONFIG.getPreferredName());
+        builder.field(TOKENS_FREQ_RATIO_THRESHOLD.getPreferredName(), tokensFreqRatioThreshold);
+        builder.field(TOKENS_WEIGHT_THRESHOLD.getPreferredName(), tokensWeightThreshold);
+        if (onlyScorePrunedTokens) {
+            builder.field(ONLY_SCORE_PRUNED_TOKENS_FIELD.getPreferredName(), onlyScorePrunedTokens);
```

Comment:
I can see the value in doing it both ways. We'll never change the default value here though and this is only available for the `rescore` method so it might be confusing to leverage an optional feature like this.

Reply:
If we're 100% sure we'll never change the default value for this, it's fine to leave it as is

Reply:
`onlyScorePrunedTokens` should be written to `XContent` and parsed. What about `async search`? 

What about anything that stores a query or rescorer in a config (ltr model?)?

Reply:
@benwtrent We can do this if we need to, but given the fact that we don't really want to change this most of the time, is this OK to hide with sane defaults like we do here? CC: @jimczi for input.